File: sdk/js/src/transport.ts
/**
 * HTTP transport for the aria-agent-cloud **beta Agents** API.
 *
 * Only `fetch` is used, so the SDK runs on Node 18+, Bun, Deno and edge runtimes.
 */

import type { ClientOptions } from "./types.js";

export const DEFAULT_BETA_HEADER = "agents=v1";

export interface ResolvedClient {
  baseUrl: string;
  apiKey?: string;
  betaHeader: string;
  fetch: typeof globalThis.fetch;
}

export function resolveClient(options: ClientOptions = {}): ResolvedClient {
  const fetchImpl =
    options.fetch ??
    (globalThis as unknown as { fetch?: typeof globalThis.fetch }).fetch;
  if (!fetchImpl) {
    throw new Error("no fetch implementation available; pass `client.fetch`");
  }
  return {
    baseUrl: (options.baseUrl ?? process.env.ARIA_AGENT_BASE_URL ?? "http://localhost:3000").replace(
      /\/+$/,
      "",
    ),
    apiKey: options.apiKey ?? process.env.ARIA_AGENT_API_KEY ?? undefined,
    betaHeader: options.betaHeader ?? DEFAULT_BETA_HEADER,
    fetch: fetchImpl,
  };
}

function headers(client: ResolvedClient, extra: Record<string, string> = {}): Record<string, string> {
  const h: Record<string, string> = {
    "OpenAI-Beta": client.betaHeader,
    ...extra,
  };
  if (client.apiKey) {
    h["Authorization"] = `Bearer ${client.apiKey}`;
  }
  return h;
}

/** POST JSON and decode a JSON response. Throws on non-2xx. */
export async function postJson<T>(
  client: ResolvedClient,
  path: string,
  body: unknown,
): Promise<T> {
  const res = await client.fetch(`${client.baseUrl}${path}`, {
    method: "POST",
    headers: headers(client, { "Content-Type": "application/json" }),
    body: JSON.stringify(body),
  });
  if (!res.ok) {
    throw new Error(`aria agent request failed (${res.status}): ${await safeText(res)}`);
  }
  return (await res.json()) as T;
}

/** GET JSON. Throws on non-2xx. */
export async function getJson<T>(client: ResolvedClient, path: string): Promise<T> {
  const res = await client.fetch(`${client.baseUrl}${path}`, {
    method: "GET",
    headers: headers(client),
  });
  if (!res.ok) {
    throw new Error(`aria agent request failed (${res.status}): ${await safeText(res)}`);
  }
  return (await res.json()) as T;
}

/** POST JSON and stream the SSE response body, yielding decoded `data:` frames. */
export async function* postJsonStream(
  client: ResolvedClient,
  path: string,
  body: unknown,
): AsyncGenerator<Record<string, any>> {
  const res = await client.fetch(`${client.baseUrl}${path}`, {
    method: "POST",
    headers: headers(client, {
      "Content-Type": "application/json",
      Accept: "text/event-stream",
    }),
    body: JSON.stringify(body),
  });
  if (!res.ok || !res.body) {
    throw new Error(`aria agent stream failed (${res.status}): ${await safeText(res)}`);
  }
  yield* readSse(res.body);
}

/**
 * Decode an SSE body into JSON frames.
 *
 * The aria envelope has **no** `data: [DONE]` sentinel: the terminal frame is
 * `agent.turn.completed` / `agent.turn.failed`, so the stream simply ends when
 * the server closes it.
 */
export async function* readSse(
  body: AsyncIterable<Uint8Array | string>,
): AsyncGenerator<Record<string, any>> {
  const decoder = new TextDecoder();
  let buffer = "";
  for await (const chunk of body as AsyncIterable<any>) {
    buffer += typeof chunk === "string" ? chunk : decoder.decode(chunk, { stream: true });
    let idx: number;
    while ((idx = buffer.indexOf("\n\n")) !== -1) {
      const raw = buffer.slice(0, idx);
      buffer = buffer.slice(idx + 2);
      for (const line of raw.split("\n")) {
        if (!line.startsWith("data:")) continue;
        const data = line.slice(5).trim();
        if (!data || data === "[DONE]") continue;
        try {
          yield JSON.parse(data);
        } catch {
          // Ignore keep-alives / malformed frames.
        }
      }
    }
  }
}

async function safeText(res: { text?: () => Promise<string> }): Promise<string> {
  try {
    return res.text ? await res.text() : "";
  } catch {
    return "";
  }
}

File: sdk/python/ariacompute_agent/transport.py
"""HTTP transport for the aria-agent-cloud **beta Agents** API.

Only the standard library is used (``urllib``), so the package installs without
any runtime dependency and works in restricted environments.
"""

from __future__ import annotations

import json
import os
import urllib.error
import urllib.request
from typing import Any, AsyncIterator, Iterator, Optional

from .types import ClientOptions

DEFAULT_BETA_HEADER = "agents=v1"


class AgentTransportError(RuntimeError):
    """Raised when the cloud returns a non-2xx response."""


def resolve_client(options: Optional[ClientOptions] = None) -> ClientOptions:
    """Fill in defaults from the environment."""
    options = options or ClientOptions()
    base_url = options.base_url or os.environ.get("ARIA_AGENT_BASE_URL") or "http://localhost:3000"
    api_key = options.api_key or os.environ.get("ARIA_AGENT_API_KEY")
    return ClientOptions(
        base_url=base_url.rstrip("/"),
        api_key=api_key,
        beta_header=options.beta_header or DEFAULT_BETA_HEADER,
    )


def _request(
    client: ClientOptions,
    method: str,
    path: str,
    body: Optional[dict[str, Any]] = None,
    stream: bool = False,
) -> urllib.response.addinfourl:
    data = json.dumps(body).encode("utf-8") if body is not None else None
    headers = {
        "OpenAI-Beta": client.beta_header,
        "Accept": "text/event-stream" if stream else "application/json",
    }
    if data is not None:
        headers["Content-Type"] = "application/json"
    if client.api_key:
        headers["Authorization"] = f"Bearer {client.api_key}"
    req = urllib.request.Request(  # noqa: S310 - base_url is operator-configured
        f"{client.base_url}{path}", data=data, headers=headers, method=method
    )
    try:
        return urllib.request.urlopen(req)  # noqa: S310 - see above
    except urllib.error.HTTPError as e:  # pragma: no cover - exercised via tests
        detail = e.read().decode("utf-8", "replace")[:400]
        raise AgentTransportError(f"aria agent request failed ({e.code}): {detail}") from e


def post_json(client: ClientOptions, path: str, body: dict[str, Any]) -> dict[str, Any]:
    """POST JSON and decode the JSON response."""
    with _request(client, "POST", path, body) as res:
        raw = res.read().decode("utf-8")
    return json.loads(raw) if raw else {}


def get_json(client: ClientOptions, path: str) -> dict[str, Any]:
    """GET and decode the JSON response."""
    with _request(client, "GET", path) as res:
        raw = res.read().decode("utf-8")
    return json.loads(raw) if raw else {}


def _iter_bytes(res: Any) -> Iterator[bytes]:
    while True:
        chunk = res.read(1024)
        if not chunk:
            break
        yield chunk


def iter_sse(res: Any) -> Iterator[dict[str, Any]]:
    """Decode an SSE body into JSON frames.

    The aria envelope has **no** ``data: [DONE]`` sentinel: the terminal frame
    is ``agent.turn.completed`` / ``agent.turn.failed``, so the stream simply
    ends when the server closes it.
    """
    buffer = ""
    for chunk in _iter_bytes(res):
        buffer += chunk.decode("utf-8", "replace")
        while "\n\n" in buffer:
            raw, buffer = buffer.split("\n\n", 1)
            for line in raw.splitlines():
                if not line.startswith("data:"):
                    continue
                data = line[5:].strip()
                if not data or data == "[DONE]":
                    continue
                try:
                    yield json.loads(data)
                except json.JSONDecodeError:
                    # Ignore keep-alives / malformed frames.
                    continue


def stream_json(
    client: ClientOptions, path: str, body: dict[str, Any]
) -> Iterator[dict[str, Any]]:
    """POST JSON and yield decoded SSE frames (blocking iterator)."""
    res = _request(client, "POST", path, body, stream=True)
    try:
        yield from iter_sse(res)
    finally:
        res.close()


async def async_stream_json(
    client: ClientOptions, path: str, body: dict[str, Any]
) -> AsyncIterator[dict[str, Any]]:
    """Async wrapper over :func:`stream_json` (the transport is blocking I/O)."""
    for frame in stream_json(client, path, body):
        yield frame
